### PR TITLE
Organizing Farm Data with farmOS Post

### PR DESCRIPTION
**Pull request by symbioquine**
A post discussing some of the high-level considerations when evaluating what data to store in farmOS and how to organize it.

### CODE REVIEW

File: posts/2023/organizing-farm-data/index.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+> the weights might be recorded as an animal count, total weight, and average weight for the whole group or as individual weight logs for each animal. One thing that
+> can help with that decision is determining whether there are other ways to re-use the data. e.g. Would having individual animal pasture locations and weight history
+> be helpful in finding animals or making care decisions day-to-day?
+> 
+> farmOS is very unopinionated about how one uses its data model. This can be helpful but also means we need to exercise some creativity when choosing the best fit for
+> any given use-case. In the employee tool and material finding example above, it might be tempting to track the locations of each tool or material when employees move them
+> around, but it might not be realistic to expect employees to be updating those locations very well in real-time. Instead, you could choose to just model the normal storage
+> location for the tool or material assets that way employees know where to look first and where to put the materials back when they're done with them. By also referencing
```
Comment (mstenta) [inline posts/2023/organizing-farm-data/index.md]: This reads a bit awkward. Maybe just missing punctuation between "material assets" and "that way"?

File: posts/2023/organizing-farm-data/index.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+> It may be obvious, but the first thing to consider is what you are trying to achieve and how is that different from where you are now.
+> 
+> It can be tempting to blindly pursue a comprehensive record system, record ad-hoc data without a plan, or even skip record keeping altogether. Chances are that none of
+> those is a great fit for your farm. Instead, most farms probably need something in the middle of those extremes. You can't record every minutiae since it would
+> take too much time and have diminishing returns. Completely ad-hoc data entry probably will not serve any particular purpose well, but skipping record keeping
+> entirely would be giving up on a valuable mechanism to keep things running smoothly and improve operations over time.
+> 
+> What to do?
+> 
+> *Note: The thought process outlined here is almost certainly not original, but is hopefully presented it in a way that is tailored to be immediately useful to those entering data in farmOS.*
```
Comment (mstenta) [inline posts/2023/organizing-farm-data/index.md]: Not sure this is necessary, but fine to include.

File: posts/2023/organizing-farm-data/index.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+> 
+> ## Developing Conventions
+> 
+> The patterns of what data gets entered and how it is recorded/used are called [conventions](https://farmos.org/model/convention/). It is expected that farmOS will
+> eventually have tools that can work directly with conventions, but for now conventions are a way to capture those patterns and possibly refine them over time. The
+> audience for your convention(s) is yourself and others who will use farmOS with you.
+> 
+> Basically, a convention is just a written description of the purpose of the data being recorded and how to record it. This can be a plain english (or language of your
+> choice) set of instructions for what data to enter in sufficient detail that someone who is generally familiar with farmOS, but not with your use-case can enter or
+> validate the data.
```
Comment (mstenta) [inline posts/2023/organizing-farm-data/index.md]: The second sentence might be reworked a bit to be easier to read.

File: posts/2023/organizing-farm-data/index.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+> 
+> Writing conventions - especially more detailed/formal ones - is optional, but is a good way to capture the work you have done thinking through the what, how, and why
+> of the data you plan to record. It also could be a good way to share your plan with collaborators or employees.
+> 
+> ## Next Steps
+> 
+> Hopefully this has been a helpful overview of one strategy for choosing what data to record in farmOS. Ideally, it will inspire you to choose something strategic to
+> improve through better record keeping - and start soon.
+> 
+> Remember not to get too wrapped up in making it perfect initially. It can help to set up a rhythm for improvements to your record keeping - possibly yearly/seasonal -
```
Comment (mstenta) [inline posts/2023/organizing-farm-data/index.md]: s/seasonal/seasonally? for consistency with "yearly"

**Review by mstenta — DISMISSED** [review]
This is great @symbioquine! I just left a couple of minor comments/suggestions, but this could be merged as-is too.

**Comment by mstenta** [comment]
Oops my review conflicted with your force push. Hopefully the comments still make sense. :sweat_smile:

**Comment by symbioquine** [comment]
> Oops my review conflicted with your force push. Hopefully the comments still make sense. sweat_smile

Yeah, I accidentally checked in the version that I posted to the forum where all lines were preceded with "> ".

File: posts/2023/organizing-farm-data/index.md (hunk 1 above)
Comment (symbioquine) [reply posts/2023/organizing-farm-data/index.md]: Good catch!

**Comment by gbathree** [comment]
I like the initial setup... it does seem obvious, but it's very often overlooked, to only keep data that actually matters.  I feel like I have to say and repeat that a lot.  I also like how you lay out how to determine what to keep - I feel like our needs assessment process with the farm support group asked similar questions, and probably we'd benefit from something of a rubric around this.

"Data Structure and Re-usability" is really where the problem is met.  I think you lay it out well, but from a techno-farmer perspective, someone who enjoys solving the techno-parts of farming.  I can say from talking with many people about FarmOS, this is where FarmOS falls flat... very few people see it as their role to figure this out.  It's like telling someone 'now design and 3D print this part' when they are just trying to fix their bathroom towel hanger (an argument the 3D printer community made for many many years also, thinking they were solving a problem)... it's only possible / enjoyable for a very small section of people.  For everyone else, they would reasonable ask 'why didn't you design it for me so it just works'.

The core issue here is that we're giving offering a problem to solve a problem, and I think that is the crux of the DIY mindset.  DIY is great when people enjoy the D part... but when they don't, or their too busy, etc then it's as if someone is giving you sand and calling it a mirror.

Sorry - I'm done sounding negative here!  I think you laid the current solution out well, but (depending on the audience) the final outcome may fall flat.  Maybe spend more time at the end providing a broader vision than just "we're solving a problem with a problem" oriented towards the non-techie farmer.  Specifically, through collaboration and transparency and sharing and all that jazz, we are actually in the process of solving a problem with a very robust and community built _solution_.

I'd also be happy to follow up on this on the blog if those are parts you'd feel would be better in a separate blog post.  The funny thing is even the work we've done w/ schema conventions is just deeper down the techie farmer rabbithole :)  So we have to point to that far future, where you can choose an opinionated farmOS instance which follows a convention, and where 'it just works' but it also retains that flexibility to account for your specific case.  That's the sweet spot!

**Comment by symbioquine** [comment]
100% @gbathree! I've been stewing over that exact dilemma for a while. (Who can we honestly pitch farmOS to right now as-is?)

> Sorry - I'm done sounding negative here!

Not at all.

> I think you laid the current solution out well, but (depending on the audience) the final outcome may fall flat. Maybe spend more time at the end providing a broader vision than just "we're solving a problem with a problem" oriented towards the non-techie farmer. Specifically, through collaboration and transparency and sharing and all that jazz, we are actually in the process of solving a problem with a very robust and community built _solution_.

I want to speak to what is actionable right now and end with a call-to-action that is more self-serving for the new user (in a positive way) - rather than just the aspirational "help us build a community that will eventually give you [the non-techie farmer] an opinionated turn-key solution." There will necessarily be some self-selection in terms of who that call-to-action resonates with.

> I think you lay it out well, but from a techno-farmer perspective, someone who enjoys solving the techno-parts of farming. I can say from talking with many people about FarmOS, this is where FarmOS falls flat... very few people see it as their role to figure this out.

This is a good data-point and reminder. I tend to assume that many farmers are somewhat intuitive systems thinkers and end up doing process design a lot already (IRL). If that's true, it's a relatively small step to say; "Hey, we have this structure that might fit your data/processes." Either way, I'm hoping to speak to the next wider sphere of possible farmOS users. Hopefully that includes more than just "techno-farmers", but that might be where we're at...

> I'd also be happy to follow up on this on the blog if those are parts you'd feel would be better in a separate blog post. The funny thing is even the work we've done w/ schema conventions is just deeper down the techie farmer rabbithole :) So we have to point to that far future, where you can choose an opinionated farmOS instance which follows a convention, and where 'it just works' but it also retains that flexibility to account for your specific case. That's the sweet spot!

My response above probably sounds like an oblique way of saying, "you're right, but I don't want to change what I've written." At some level that's true, but not because I don't value what you're pointing out or want to put more effort into the post. Rather, I just suspect that - in order to welcome more folks into the project - we need to offer them a way of thinking about what data can go into farmOS which provides some value to them right away. Emphasizing the aspirational part too much could also scare folks away. That said, I take your point and I'll see if I can work a bit more sense of the grander vision in.

It would also be really awesome to have a whole post that describes your vision for that "conventions endgame"! I'm happy to help with editing/proof-reading/etc.

**Comment by gbathree** [comment]
Super - yeah I think we're on the same page.  I will consider a blog post follow up as you said, but you're also right that you are giving people actionable steps today which is in some ways better than saying 'wait till we're done', even if they may be actionable only to a subset of users.  Progress continues :)

**Comment by mstenta** [comment]
>  I can say from talking with many people about FarmOS, this is where FarmOS falls flat... very few people see it as their role to figure this out.
> ...
> The core issue here is that we're giving offering a problem to solve a problem, and I think that is the crux of the DIY mindset.

Nail on the head @gbathree! This is a great discussion here - I almost wish we had some of this in the forum because it will be lost in this GitHub PR. :-)

The origins of farmOS, like many grassroots open source projects, are DIY at their core. That's not to say they can't go on to be the foundation of complete solutions in the future! Linux itself is a great example of those possibilities - and also of separation of concerns. The Linux "kernel" doesn't necessarily care what UI is built on top of it, that's up to other projects like Gnome, KDE, etc. And even those don't care about specific use-cases like graphic design or programming, that's for projects like GIMP and Eclipse. We build best when we build in layers!

**Comment by mstenta** [comment]
(Side note @gbathree: if you "approve" this PR by going to the "Files changed" tab above, then clicking "Review changes", selecting "Approve" and clicking "Submit review" then @symbioquine will be able to merge this PR and publish it. I gave you and others permission to approve PRs in this repo recently. Helps to disperse the effort to get these things reviewed and merged among others in the community!)

**Comment by symbioquine** [comment]
> (Side note @gbathree: if you "approve" this PR by going to the "Files changed" tab above, then clicking "Review changes", selecting "Approve" and clicking "Submit review" then @symbioquine will be able to merge this PR and publish it. I gave you and others permission to approve PRs in this repo recently. Helps to disperse the effort to get these things reviewed and merged among others in the community!)

Good for future reference, but I should note that approval isn't the blocker here yet. I'm still planning to update this PR - to address the feedback you and Greg have given - and I don't think the approval will persist if I add/amend the commits of the PR.

**Comment by buzzkc** [comment]
This document is a very good start for someone that is wanting to really jump in and track specific goals or from a researchers perspective. From someone that has been trying to set farmOS up for my own farm. I've struggled a bit with organization of the initial structure, and given we have no livestock currently, I've been trying to figure out how farmOS would fit into an agroforestry (trees, plantings, harvest, nut harvests, syrup collection, etc). 
I have also been thinking about it from my neighbors point of view for managing his livestock and what he'd need to track his assets, create logs, manage inventory, review reports, etc.  I don't think I could sell him on it from a new instance standpoint, but perhaps adding a few step-by-step guides on how to setup farmOS for the more basic farms out there rather than from a researchers perspective. Start with creating the location>areas>assets>>logs with explanation of how/why I'd want to setup farmOS this way. So one for Beef cattle, one for managing crops, one for the bee keeper, etc. These certainly can be community contributed  by the community and added to the list over time.

**Comment by symbioquine** [comment]
Thanks @buzzkc!

> This document is a very good start for someone that is wanting to really jump in and track specific goals or from a researchers perspective

I was hoping to target a slightly wider audience of the technically inclined farmer - not just "researchers". Like I said above, I tend to think that farmers already practice iterating on their processes/systems all the time and it's not too big of a stretch (for many) to transition to thinking about their "data processes/systems" as well.

> perhaps adding a few step-by-step guides on how to setup farmOS for the more basic farms out there rather than from a researchers perspective. Start with creating the location>areas>assets>>logs with explanation of how/why I'd want to setup farmOS this way. So one for Beef cattle, one for managing crops, one for the bee keeper, etc. These certainly can be community contributed by the community and added to the list over time.

For one, I assume you've seen https://farmos.org/blog/2022/getting-started/ I'd like there to be a whole series of posts like that which go into mapping various details of a farm's data into the general farmOS data model. (Maybe you'll write one for the aspects you're thinking through...)

Another angle is conventions. The post I was teasing here will ideally be written by @gbathree and share the larger vision of having more opinionated sets of conventions which a farmOS instance can "subscribe to". For common use-cases, many of the decisions I'm describing in this post could be captured in those conventions - along with automatic validation and tooling for entering the data according to the conventions.

**Review by mstenta — DISMISSED** [review]

**Comment by buzzkc** [comment]
@symbioquine I had not seen that blog post. I had looked over most of the farmOS.org documentation, but not the blogs. Thank you. As I mentioned on the call this morning, I went a little overboard adding modules to my instance and need to dial that back, then start again with the basics.

**Review by paul121 — APPROVED** [review]

**Review by mstenta — APPROVED** [review]

**Pull request merged**